core/src/ledger/masp_utils.rs

```diff
@@ -0,0 +1,113 @@
+//! MASP utilities
+
+use masp_primitives::merkle_tree::CommitmentTree;
+use masp_primitives::sapling::Node;
+use masp_primitives::transaction::Transaction;
+
+use super::storage_api::{StorageRead, StorageWrite};
+use crate::ledger::storage_api::{Error, Result};
+use crate::types::address::MASP;
+use crate::types::hash::Hash;
+use crate::types::storage::{BlockHeight, Epoch, Key, KeySeg, TxIndex};
+use crate::types::token::{
+    Transfer, HEAD_TX_KEY, MASP_NOTE_COMMITMENT_TREE_KEY,
+    MASP_NULLIFIERS_KEY_PREFIX, PIN_KEY_PREFIX, TX_KEY_PREFIX,
+};
+
+// Writes the nullifiers of the provided masp transaction to storage
+fn reveal_nullifiers(
+    ctx: &mut impl StorageWrite,
+    transaction: &Transaction,
+) -> Result<()> {
+    for description in transaction
+        .sapling_bundle()
+        .map_or(&vec![], |description| &description.shielded_spends)
+    {
+        let nullifier_key = Key::from(MASP.to_db_key())
+            .push(&MASP_NULLIFIERS_KEY_PREFIX.to_owned())
+            .expect("Cannot obtain a storage key")
+            .push(&Hash(description.nullifier.0))
+            .expect("Cannot obtain a storage key");
+        ctx.write(&nullifier_key, ())?;
+    }
+
+    Ok(())
+}
+
+/// Appends the note commitments of the provided transaction to the merkle tree
+/// and updates the anchor
+/// NOTE: this function is public as a temporary workaround because of an issue
```

Comment:
So initially this was compiled and called directly from WASM. Unfortunately I found the the wasm context was crashing constantly when calling the `append` method of the tree. At the same time, though, the MASP native VP had no problem running the same function. I even tried ignoring the error coming out of the `append` method but nothing changed (expected because that error is only triggered when the tree is full which is not our case). I did some tests and found that the issue manifests when the `append` function calls the `combine` method of `Node`, so everything works fine as long as one of the left or right field of the tree are empty but crashes otherwise. 

So as a workaround, this function is now called separately and as a host function if we are in the wasm `transfer`,

Reply:
This is fine. We can investigate the causes of this failure in the future for curiosity's sake.